sys/stdio_nimble/stdio_nimble.c

```diff
@@ -0,0 +1,399 @@
+        struct ble_gatt_access_ctxt *ctxt, void *arg);
+
+void debug(const void* buffer, size_t len) {
+    uart_write(STDIO_UART_DEV, (const uint8_t *)buffer, len);
+}
+
+/**
+ * @brief Dummy access callback, because nimble requires one
+ */
+static int gatt_svr_chr_access_noop(
+        uint16_t conn_handle, uint16_t attr_handle,
+        struct ble_gatt_access_ctxt *ctxt, void *arg) {
+
+    (void) conn_handle;
+    (void) attr_handle;
+    (void) ctxt;
+    (void) arg;
+
+    return 0;
+}
+
+/**
+ * @brief Struct to define the stdio bluetooth service with its characteristics
+ */
+static const struct ble_gatt_svc_def _gatt_svr_svcs[] =
+{
+    /*
+     * access_cb defines a callback for read and write access events on
+     * given characteristics
+     */
+    {
+        /* Service: stdio */
+        .type = BLE_GATT_SVC_TYPE_PRIMARY,
+        .uuid = (ble_uuid_t*) &gatt_svr_svc_stdio_uuid.u,
+        .characteristics = (struct ble_gatt_chr_def[]) { {
+            /* Characteristic: stdout */
+            .uuid = (ble_uuid_t*) &gatt_svr_chr_stdout_uuid.u,
+            .access_cb = gatt_svr_chr_access_noop,
+            .val_handle = &_val_handle_stdout,
+            .flags = BLE_GATT_CHR_F_INDICATE,
+        }, {
+            /* Characteristic: stdin */
+            .uuid = (ble_uuid_t*) &gatt_svr_chr_stdin_uuid.u,
+            .access_cb = gatt_svr_chr_access_stdin,
+            .flags = BLE_GATT_CHR_F_WRITE,
+        }, {
+            0, /* No more characteristics in this service */
+        }, }
+    },
+    {
+        0, /* No more services */
+    },
+};
+
+static int _gap_event_cb(struct ble_gap_event *event, void *arg);
+
+static void _send_from_ringbuffer(bool mutex_locked);
+
+void stdio_init(void)
+{
+    ringbuffer_init(&_rb_stdin, _rb_stdin_mem, CONFIG_STDIO_NIMBLE_STDIN_BUFSIZE);
+    ringbuffer_init(&_rb_stdout, _rb_stdout_mem, CONFIG_STDIO_NIMBLE_STDOUT_BUFSIZE);
+
+    uart_init(STDIO_UART_DEV, STDIO_UART_BAUDRATE, NULL, NULL);
+}
+
+ssize_t stdio_read(void *buffer, size_t count)
+{
+    ssize_t res;
+
+    /* block until at least on character was read from the buffer */
+    while (!(res = ringbuffer_get(&_rb_stdin, buffer, count))) {};
+
+    return res;
+}
+
+ssize_t stdio_write(const void* buffer, size_t len)
+{
+    mutex_lock(&_rb_stdout_mutex);
+
+    unsigned int free = ringbuffer_get_free(&_rb_stdout);
+    unsigned int consumed = 0;
+
+    if (free > 0) {
+        unsigned int to_send = (free <= len) ? free : len;
+        consumed = ringbuffer_add(&_rb_stdout, buffer, to_send);
+    }
+
+    mutex_unlock(&_rb_stdout_mutex);
+
+    /* initially trigger chained async calls (chained via "_gap_event_cb") */
+    if (mutex_trylock(&_send_mutex)) {
+        if (!_sending_in_progress) {
+            _send_from_ringbuffer(true);
+        }
+        else {
+            /* there is an active call chain, we must not initiate a new one */
+            mutex_unlock(&_send_mutex);
+        }
+    }
+
+    return consumed;
```

Comment:
Since you are returning the consumed data, does that mean that non consumed data will be resubmited?

Reply:
I can't really answer this question. I guess this depends on the libraries that are used for print-functions and how resilient they are designed. When I shrank the buffer to 4 bytes then I only got 4 bytes (`"4 bytes sent successfully"`) so at least the libraries involved in this PR and the `nrf52dk` don't seem to care about the return value of `stdio_write`.